src/main/java/com/netgrif/application/engine/pdf/generator/service/PdfDataHelper.java

```diff
@@ -98,16 +97,31 @@ public void generatePdfFields() {
         this.dataGroups = dataService.getDataGroups(taskId, resource.getTextLocale()).getData();
 
         dataGroups.forEach(dataGroup -> {
-            refreshGrid(dataGroup);
-            dataGroup.getFields().getContent().forEach(field -> {
-                        generateField(dataGroup, field);
-                    }
-            );
-            this.lastX = Integer.MAX_VALUE;
+            if (dataGroup.getParentTaskRefId() == null) {
+                refreshGrid(dataGroup);
+                generateFromDataGroup(dataGroup);
+                this.lastX = Integer.MAX_VALUE;
+            }
         });
         Collections.sort(pdfFields);
     }
 
+    private void generateFromDataGroup(DataGroup dataGroup) {
+        Collection<LocalisedField> fields = dataGroup.getFields().getContent();
+        if (dataGroup.getLayout() != null && dataGroup.getLayout().getType() != null && dataGroup.getLayout().getType().equals("grid")) {
+            fields = fields.stream().sorted(Comparator.<LocalisedField, Integer>comparing(f -> f.getLayout().getY()).thenComparing(f -> f.getLayout().getX())).collect(Collectors.toList());
+        }
+       fields.forEach(field -> {
+                    if (field.getType().equals(FieldType.TASK_REF)) {
+                        Optional<DataGroup> taskRefGroup = this.dataGroups.stream().filter(dg -> dg.getParentTaskRefId() != null && dg.getParentTaskRefId().equals(field.getStringId())).findFirst();
```

Comment:
The exprassion `dg.getParentTaskRefId() != null && dg.getParentTaskRefId().equals(field.getStringId())` in lambda function could be replaced with `Objects.equals(dg.getParentTaskRefId(), field.getStringId())` .

Reply:
Done